### PR TITLE
Enrich MCX requests with Token headerIDSDEV-638

### PR DESCRIPTION
**Pull request by seshangamage**

### CODE REVIEW

File: components/api-invocation-hanlder/.gitignore [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+*.iml
+*.iws
+*.releaseBackup
+*.ipr
+.idea
+*pom.xml.next
+*pom.xml.tag
+*release.properties
+features/com.wso2telco.ids.dashboard.ui.feature/src/main/resources/generated/dashboard.zip
+*pom.xml.swp
```
Comment (Buddhima) [inline components/api-invocation-hanlder/.gitignore]: This file is not needed

File: components/api-invocation-hanlder/pom.xml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    <parent>
+        <artifactId>component-dep</artifactId>
+        <groupId>com.wso2telco.dep</groupId>
+        <version>2.2.8-SNAPSHOT</version>
+        <relativePath>../../pom.xml</relativePath>
+    </parent>
+
+    <groupId>com.wso2telco.dep</groupId>
+    <artifactId>api-invocation-handler</artifactId>
+    <version>1.0.0-SNAPSHOT</version>
```
Comment (Buddhima) [inline components/api-invocation-hanlder/pom.xml]: Avoid using separate versions inside components. Allow to inherit from parent

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    private String getClientIdForAuthorizeCall(String fullPath) {
+        String clientId = null;
+        if (fullPath.contains(AUTH_ENDPOINT)) {
+            if (log.isDebugEnabled()) {
+                log.debug(AUTH_ENDPOINT);
+            }
+            clientId = fullPath.split("client_id=")[1].split("&")[0];
+
```
Comment (Buddhima) [inline components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java]: remove additional spaces

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    private String getClientId(String fullPath, Map headerMap, MessageContext messageContext) {
+        String clientId = null;
+        if (fullPath.contains(TOKEN_ENDPOINT) || fullPath.contains(USERINFO_ENDPOINT)) {
+            String basicAuth = (String) headerMap.get(AUTH_HEADER);
+            headerMap.remove(AUTH_HEADER);
+            messageContext.setProperty(TEMP_AUTH_HEADER, basicAuth);
+            clientId = getClientIdForTokenCall(fullPath, basicAuth);
+            if (clientId == null)
+                getClientIdForUserInfoCall(fullPath, basicAuth);
+
```
Comment (Buddhima) [inline components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java]: additional spaces...

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            }
+            byte[] valueDecoded = Base64.decodeBase64(basicAuth.split(" ")[1].getBytes());
+            String decodeString = new String(valueDecoded);
+            clientId = decodeString.split(":")[0];
+        }
+        return clientId;
+    }
+
+    private boolean processResponse(String clientId, Map headerMap, String fullPath) {
+
```
Comment (Buddhima) [inline components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java]: additional spaces ...

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        if (clientId == null)
+            return false;
+
+        else if (fullPath.contains(USERINFO_ENDPOINT)) {
+            headerMap.put(AUTH_HEADER, TOKEN_TYPE + clientId);
+            return true;
+
```
Comment (Buddhima) [inline components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java]: additional spaces ...

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        }
+    }
+
+    @Override
+    public boolean handleResponse(MessageContext messageContext) {
+        return true;
+    }
+
+    private void callTokenPool(final AddNewSpDTO newSpDto) {
+        ExecutorService executorService = Executors.newFixedThreadPool(1);
```
Comment (Buddhima) [inline components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java]: use executor service effectively. This gets executed per request.

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/dto/AddNewSpDTO.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+package com.wso2telco.dep.apihandler.dto;
+import java.util.List;
+
+public class AddNewSpDTO {
```
Comment (Buddhima) [inline components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/dto/AddNewSpDTO.java]: use Nouns for class names

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/dto/AddNewSpDTO.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+    private String ownerId;
+    private static final String TOKEN_URL = "https://localhost:8243/token";
+    private static final long DEFAULT_CONNECTION_RESET_TIME = 4000;
+    private static final int RETRY_ATTEMPT = 3;
+    private static final int RETRYMAX = 10;
+    private static final int RETRYDELAY = 20000;
```
Comment (Buddhima) [inline components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/dto/AddNewSpDTO.java]: can't we use these as public constants?

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/internal/ApiInvocationHandlerServiceComponent.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+ ******************************************************************************/
+package com.wso2telco.dep.apihandler.internal;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.synapse.rest.AbstractHandler;
+import org.osgi.service.component.ComponentContext;
+
+import com.wso2telco.dep.apihandler.ApiInvocationHandler;
+
+public class ApiInvocationHandlerServiceComponent {
```
Comment (Buddhima) [inline components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/internal/ApiInvocationHandlerServiceComponent.java]: Any use of this class?

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/util/ReadPropertyFile.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+                try {
+                    input.close();
+                } catch (IOException e) {
+                    log.error(e.getStackTrace());
+                }
+            }
+        }
+
+        return propertyMap;
+
```
Comment (Buddhima) [inline components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/util/ReadPropertyFile.java]: remove empty lines

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/util/TokenPoolUtil.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+import java.io.IOException;
+import java.util.Map;
+
+public class TokenPoolUtil {
+
+    private static final Log log = LogFactory.getLog(TokenPoolUtil.class);
+    private static final String APPLICATION_JSON = "application/json";
+    private static final String TOKEN_POOL_URL = "token_pool_url";
+
+    private TokenPoolUtil() {
```
Comment (Buddhima) [inline components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/util/TokenPoolUtil.java]: remove

File: components/api-invocation-hanlder/.gitignore (hunk 1 above)
Comment (seshangamage) [reply components/api-invocation-hanlder/.gitignore]: Removed

File: components/api-invocation-hanlder/pom.xml (hunk 2 above)
Comment (seshangamage) [reply components/api-invocation-hanlder/pom.xml]: Removed from POM

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java (hunk 3 above)
Comment (seshangamage) [reply components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java]: done

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java (hunk 4 above)
Comment (seshangamage) [reply components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java]: done

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java (hunk 5 above)
Comment (seshangamage) [reply components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java]: done

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java (hunk 6 above)
Comment (seshangamage) [reply components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java]: done

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java (hunk 7 above)
Comment (seshangamage) [reply components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/ApiInvocationHandler.java]: This Method is only trigger if Token pool is enabled as per requirement
 if (propertyMap.get(TOKEN_POOL_ENABLED).equals("true")) {
            if (log.isDebugEnabled()) {
                log.debug("TOken Pool Service Enabled for Insertion");
            }
            AddNewSpDTO newSpDto = new AddNewSpDTO();
            ArrayList<TokenDTO> tokenList = new ArrayList<TokenDTO>();
            newSpDto.setOwnerId(clientId);
            tokenDTO.setTokenAuth(getbase64EncodedTokenAouth(tokenDTO.getTokenAuth()));
            tokenList.add(tokenDTO);
            newSpDto.setSpTokenList(tokenList);
            callTokenPool(newSpDto);
        }

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/dto/AddNewSpDTO.java (hunk 8 above)
Comment (seshangamage) [reply components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/dto/AddNewSpDTO.java]: Re factored  to NewSpDTO

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/internal/ApiInvocationHandlerServiceComponent.java (hunk 10 above)
Comment (seshangamage) [reply components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/internal/ApiInvocationHandlerServiceComponent.java]: This class was not required

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/util/ReadPropertyFile.java (hunk 11 above)
Comment (seshangamage) [reply components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/util/ReadPropertyFile.java]: done

File: components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/util/TokenPoolUtil.java (hunk 12 above)
Comment (seshangamage) [reply components/api-invocation-hanlder/src/main/java/com/wso2telco/dep/apihandler/util/TokenPoolUtil.java]: done

**Pull request closed without merging**